src/backends/torch/native/templates/ttransformer.cc

```diff
@@ -0,0 +1,131 @@
+                  std::to_string(_embed_dropout));
+    _embedder = register_module("embedder",
+                                Embedder(_input_dim, _input_len, _embed_type,
+                                         _embed_activation, _embed_dim,
+                                         _embed_nlayers, _embed_dropout));
+
+    _logger->info("pe type: {}  dropout: {}", std::to_string(_petype),
+                  std::to_string(_pe_dropout), std::to_string(_embed_dropout));
+    _pe = register_module("pe", PositionalEncoding(_input_len, _embed_dim,
+                                                   _petype, _pe_dropout));
+
+    _logger->info(
+        "encoder nheads: {}  nlayers: {}  hidden_dim: {}  dropout: {}",
+        std::to_string(_encoder_nheads), std::to_string(_encoder_nlayers),
+        std::to_string(_encoder_hidden_dim), std::to_string(_encoder_dropout));
+    _encoder = register_module(
+        "TEncoder",
+        TEncoder(_input_len, _embed_dim, _encoder_nheads, _encoder_nlayers,
+                 _encoder_hidden_dim, _encoder_dropout, _encoder_activation));
+
+    if (_simple_decoder)
+      _logger->info("decoder mlp: nlayers: {}  hidden_dim: {}  dropout: {}",
+                    std::to_string(_decoder_nlayers),
+                    std::to_string(_decoder_hidden_dim),
+                    std::to_string(_encoder_dropout));
+    else
+      _logger->info(
+          "decoder transformer: nheads: {} nlayers: {}  hidden_dim: {}  "
+          "dropout: {}",
+          std::to_string(_decoder_nheads), std::to_string(_decoder_nlayers),
+          std::to_string(_decoder_nlayers),
+          std::to_string(_decoder_hidden_dim),
+          std::to_string(_encoder_dropout));
+    _decoder = register_module(
+        "TDecoder",
+        TDecoder(_simple_decoder, _embed_dim, _input_len, _output_dim,
+                 _output_len, _decoder_nheads, _decoder_nlayers,
+                 _decoder_hidden_dim, _decoder_dropout));
+    if (!_autoreg)
+      _encoder_mask
+          = register_buffer("encoder_mask", generate_mask_none(_input_len));
+    else
+      {
+        _encoder_mask
+            = register_buffer("encoder_mask", generate_mask_past(_input_len));
+        _decoder_mask
+            = register_buffer("decoder_mask", generate_mask_past(_output_len));
+      }
+
+    if (_embed_dim % _encoder_nheads != 0)
+      {
+        throw MLLibBadParamException(
+            "embed_dim must be divisible by encoder_nheads");
+      }
+  }
+
+  torch::Tensor TTransformer::generate_mask_none(int len)
+  {
+    return torch::ones({ len, len }).to(torch::kFloat32);
+  }
+  torch::Tensor TTransformer::generate_mask_past(int len)
+  {
+    // TODO : past mask
+    return torch::ones({ len, len }).to(torch::kFloat32);
+  }
+
+  void TTransformer::set_encoder_mask(torch::Tensor mask)
+  {
+    // TODO : change dims
+    _encoder_mask = mask;
+  }
+
+  void TTransformer::set_decoder_mask(torch::Tensor mask)
+  {
+    // TODO : change dims
```

Comment:
Does something need to be done now ?

Reply:
decoder mask is not implemented yet, because decoder part is not tested ATM, will do in some other PR